### PR TITLE
Url hash progress

### LINKED ISSUE #32: Use url hashes to track story progress, and let users start a story in the middle / end
You should be able to share the end result of a story. Let's do this by adding a hash url component with push state:

https://storyboard.viget.com/find-out-scholarships-available#final-screen

This could get messy with modifiers/items, could consider a way to encrypt the item/stats state so users can't just give themselves a million health points

### PR DESCRIPTION
**Pull request by shanexavier313**
# Summary
Fixes #32. Adds a url hash which tracks progress through a story so that users can visit the url and pop in to a story mid-progress. 

## Logic
- In `componentDidUpdate()` (i.e. every time there's a state change) takes `this.state`, encrypts it and sets it as a url hash
- In `componentDidMount()`, if the url has a hash value present, decrypt it and set it as the currentState instead of starting at the beginning of the story

## Comments
I'm storing the passphrase for the encryption in a private member variable of the `Player` class. Not sure if that's the best way to hide it from the user...

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -31,3 +31,5 @@
 yarn-debug.log*
 .yarn-integrity
 .vscode/
+
+package-lock.json
```
Comment (efatsi) [inline .gitignore:35]: i don't think this is wise. keeping this lock file in the git history ensures that everyone is working off of the same dependency versions

File: app/javascript/src/components/Editor.tsx [hunk 2]
```diff
@@ -72,6 +72,8 @@ class Editor extends React.Component<EditorProps, EditorState> {
     if (JSON.stringify(story) !== JSON.stringify(newStory)) {
       this.updateStory(newStory)
       this.forceUpdate()
+    } else {
+      console.log('nothing new')
```
Comment (efatsi) [inline app/javascript/src/components/Editor.tsx:76]: this seems a little out of place in this PR. Possible that you were mixing up your branches while working on url hashes + undo/redo?

Comment (shanexavier313) [reply app/javascript/src/components/Editor.tsx:76]: Yeah I accidentally forgot to switch branches from `undo-redo` when initially working on this. I stashed the work, reset `undo-redo` and checked out `url-hash-progress` from `master` and then applied the stash. Unfortunately some of the history tracking logic in `Editor` got carried over (I must have messed something up) so I deleted and added that line. I'll try resetting just `Editor` so there are no changes pushed on this branch

Comment (efatsi) [reply app/javascript/src/components/Editor.tsx:76]: Gotchya, yeah a couple tips I've got for you which might help you keep organized around that stuff a little more:
- make the branch name visible in your terminal (or somewhere prominent), which will make it easier for you to have a sense for what branch you're on. Which will increase the likelihood that you keep your work more isolated / organized
  - if you're on vanilla bash, here's a bit of .bash_profile goodness: 
    ```bash
    display_git_branch() {
      branch=$(git branch 2> /dev/null | sed -e '/^[^*]/d' -e 's/* \(.*\)/\1/')
      if [ "$branch" != "" ]; then
        echo "($branch)"
      fi
    }
    export PS1="\[\e[1;31m\]BASH:\W\$(display_git_branch) $ \[\e[0m\]"
    ```
- use [gitx](http://rowanj.github.io/gitx/) when assembling together commits. I'm normally not a fan of visual interfaces for dealing with programming stuff, except for when it comes to gitx.

File: app/javascript/src/components/Player.tsx [hunk 3]
```diff
@@ -69,15 +73,31 @@ class Player extends React.Component<PlayerProps, PlayerState> {
   }
 
   componentDidMount() {
-    this.setState({
-      focus: this.findStartKey()
-    })
+    if (window.location.hash == '') {
```
Comment (dlederle) [inline app/javascript/src/components/Player.tsx:76]: Could it also be undefined?

**Comment by efatsi** [comment]
Looks good to me. Merge it in! You have deploy access as well now so if you push this to origin/master a `cap production deploy` should send it on out to the world!

**Pull request merged**